# Question 3 

First of all, we import the required modules.

In [1]:
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
import statsmodels.graphics as sgr
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.collections import LineCollection
from statsmodels.stats.outliers_influence import summary_table
from statsmodels.formula.api import logit
from pandas.core import datetools

%matplotlib inline

C:\Users\leo\Anaconda3\lib\site-packages\statsmodels\compat\pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


We import the csv file.

In [2]:
crd_df = pd.read_csv('creditcard.csv') #read csv file 
crd_df.head(20) #select first 10 records

Time        V1        V2        V3        V4        V5        V6  \
0    0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1    0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2    1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3    1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4    2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
5    2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728   
6    4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708   
7    7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118   
8    7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818   
9    9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761   
10  10.0  1.449044 -1.176339  0.913860 -1.375667 -1.971383 -0.629152   
11  10.0  0.384978  0.616109 -0.874300 -0.094019  2.924584  3.317027   
12  10.0  1.249999 -1.221637  0.383930 -1.234899 -1.485419 -0.753230   
13  11.0  1.069374  0.287722  0.828613  2.712520 -0.178398  0.337544   
14  12.0 -2.791855 -0.327771  1.641750  1.767473 -0.136588  0.807596   
15  12.0 -0.752417  0.345485  2.057323 -1.468643 -1.158394 -0.077850   
16  12.0  1.103215 -0.040296  1.267332  1.289091 -0.735997  0.288069   
17  13.0 -0.436905  0.918966  0.924591 -0.727219  0.915679 -0.127867   
18  14.0 -5.401258 -5.450148  1.186305  1.736239  3.049106 -1.763406   
19  15.0  1.492936 -1.029346  0.454795 -1.438026 -1.555434 -0.720961   

          V7        V8        V9  ...         V21       V22       V23  \
0   0.239599  0.098698  0.363787  ...   -0.018307  0.277838 -0.110474   
1  -0.078803  0.085102 -0.255425  ...   -0.225775 -0.638672  0.101288   
2   0.791461  0.247676 -1.514654  ...    0.247998  0.771679  0.909412   
3   0.237609  0.377436 -1.387024  ...   -0.108300  0.005274 -0.190321   
4   0.592941 -0.270533  0.817739  ...   -0.009431  0.798278 -0.137458   
5   0.476201  0.260314 -0.568671  ...   -0.208254 -0.559825 -0.026398   
6  -0.005159  0.081213  0.464960  ...   -0.167716 -0.270710 -0.154104   
7   1.120631 -3.807864  0.615375  ...    1.943465 -1.015455  0.057504   
8   0.370145  0.851084 -0.392048  ...   -0.073425 -0.268092 -0.204233   
9   0.651583  0.069539 -0.736727  ...   -0.246914 -0.633753 -0.120794   
10 -1.423236  0.048456 -1.720408  ...   -0.009302  0.313894  0.027740   
11  0.470455  0.538247 -0.558895  ...    0.049924  0.238422  0.009130   
12 -0.689405 -0.227487 -2.094011  ...   -0.231809 -0.483285  0.084668   
13 -0.096717  0.115982 -0.221083  ...   -0.036876  0.074412 -0.071407   
14 -0.422911 -1.907107  0.755713  ...    1.151663  0.222182  1.020586   
15 -0.608581  0.003603 -0.436167  ...    0.499625  1.353650 -0.256573   
16 -0.586057  0.189380  0.782333  ...   -0.024612  0.196002  0.013802   
17  0.707642  0.087962 -0.665271  ...   -0.194796 -0.672638 -0.156858   
18 -1.559738  0.160842  1.233090  ...   -0.503600  0.984460  2.458589   
19 -1.080664 -0.053127 -1.978682  ...   -0.177650 -0.175074  0.040002   

         V24       V25       V26       V27       V28  Amount  Class  
0   0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1  -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2  -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3  -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4   0.141267 -0.206010  0.502292  0.219422  0.215153   69.99      0  
5  -0.371427 -0.232794  0.105915  0.253844  0.081080    3.67      0  
6  -0.780055  0.750137 -0.257237  0.034507  0.005168    4.99      0  
7  -0.649709 -0.415267 -0.051634 -1.206921 -1.085339   40.80      0  
8   1.011592  0.373205 -0.384157  0.011747  0.142404   93.20      0  
9  -0.385050 -0.069733  0.094199  0.246219  0.083076    3.68      0  
10  0.500512  0.251367 -0.129478  0.042850  0.016253    7.80      0  
11  0.996710 -0.767315 -0.492208  0.042472 -0.054337    9.99      0  
12  0.392831  0.161135 -0.354990  0.026416  0.0

Now we will run logistic regression, because we want to build a model for the binary outcome of fraud detection (0->False, 1->True).

The dependent variable will be the class column.

In [3]:
f = 'Class ~ Time+V1+V2+V3+V4+V5+V6+V7+V8+V9+V10+V11+V12+V13+V14+V15+V16+V17+V18+V19+V20+V21+V22+V23+V24+V25+V26+V27+V28+Amount'
logitfit = smf.logit(formula = str(f), data = crd_df).fit() #fit to our model

Optimization terminated successfully.
         Current function value: 0.003914
         Iterations 13


In [4]:
logitfit.summary() #summary of our model

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  Class   No. Observations:               284807
Model:                          Logit   Df Residuals:                   284776
Method:                           MLE   Df Model:                           30
Date:                Sun, 14 Jan 2018   Pseudo R-squ.:                  0.6922
Time:                        20:18:35   Log-Likelihood:                -1114.8
converged:                       True   LL-Null:                       -3621.2
                                        LLR p-value:                     0.000
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -8.3917      0.249    -33.652      0.000      -8.880      -7.903
Time       -3.742e-06   2.26e-06     -1.659      0.097   -8.16e-06    6.79e-07
V1             0.0960      0.042      2.264      0.024       0.013       0.179
V2             0.0094      0.058      0.161      0.872      -0.104       0.123
V3            -0.0079      0.053     -0.149      0.881      -0.112       0.096
V4             0.6986      0.074      9.454      0.000       0.554       0.843
V5             0.1295      0.067      1.944      0.052      -0.001       0.260
V6            -0.1198      0.074     -1.626      0.104      -0.264       0.025
V7            -0.0969      0.067     -1.453      0.146      -0.228       0.034
V8            -0.1739      0.030     -5.711      0.000      -0.234      -0.114
V9            -0.2843      0.111     -2.561      0.010      -0.502      -0.067
V10           -0.8176      0.097     -8.432      0.000      -1.008      -0.628
V11           -0.0621      0.081     -0.762      0.446      -0.222       0.098
V12            0.0909      0.087      1.045      0.296      -0.080       0.261
V13           -0.3312      0.082     -4.058      0.000      -0.491      -0.171
V14           -0.5571      0.062     -8.949      0.000      -0.679      -0.435
V15           -0.1141      0.086     -1.330      0.183      -0.282       0.054
V16           -0.1908      0.125     -1.526      0.127      -0.436       0.054
V17           -0.0216      0.070     -0.309      0.757      -0.159       0.116
V18           -0.0131      0.129     -0.102      0.919      -0.266       0.240
V19            0.0963      0.097      0.993      0.321      -0.094       0.286
V20           -0.4582      0.082     -5.607      0.000      -0.618      -0.298
V21            0.3898      0.060      6.494      0.000       0.272       0.507
V22            0.6297      0.134      4.707      0.000       0.367       0.892
V23           -0.0951      0.058     -1.629      0.103      -0.209       0.019
V24            0.1289      0.147      0.874      0.382      -0.160       0.418
V25           -0.0761      0.131     -0.582      0.560      -0.332       0.180
V26            0.0195      0.190      0.103      0.918      -0.352       0.392
V27           -0.8188      0.122     -6.686      0.000      -1.059      -0.579
V28           -0.2937      0.088     -3.332      0.001      -0.467      -0.121
Amount         0.0009      0.000      2.449      0.014       0.000       0.002
==============================================================================

Possibly complete quasi-separation: A fraction 0.31 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

Now we will calculate the probability of each record to be classified as fraud.

In [5]:
y_pred = logitfit.predict() #predict values of fraud (binary) classification
crd_df ['Prediction'] = (y_pred) #store to crd_df dataframe, Prediction column
crd_df.head(10) #select first 10 records

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9     ...           V22       V23       V24       V25  \
0  0.098698  0.363787     ...      0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425     ...     -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654     ...      0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024     ...      0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739     ...      0.798278 -0.137458  0.141267 -0.206010   
5  0.260314 -0.568671     ...     -0.559825 -0.026398 -0.371427 -0.232794   
6  0.081213  0.464960     ...     -0.270710 -0.154104 -0.780055  0.750137   
7 -3.807864  0.615375     ...     -1.015455  0.057504 -0.649709 -0.415267   
8  0.851084 -0.392048     ...     -0.268092 -0.204233  1.011592  0.373205   
9  0.069539 -0.736727     ...     -0.633753 -0.120794 -0.385050 -0.069733   

        V26       V27       V28  Amount  Class  Prediction  
0 -0.189115  0.133558 -0.021053  149.62      0    0.000630  
1  0.125895 -0.008983  0.014724    2.69      0    0.000200  
2 -0.139097 -0.055353 -0.059752  378.66      0    0.000350  
3 -0.221929  0.062723  0.061458  123.50      0    0.000147  
4  0.502292  0.219422  0.215153   69.99      0    0.000197  
5  0.105915  0.253844  0.081080    3.67      0    0.000145  
6 -0.257237  0.034507  0.005168    4.99      0    0.000552  
7 -0.051634 -1.206921 -1.085339   40.80      0    0.000467  
8 -0.384157  0.011747  0.142404   93.20      0    0.000237  
9  0.094199  0.246219  0.083076    3.68      0    0.000115  

[10 rows x 32 columns]

In [6]:
x = crd_df[crd_df.Class == 1 ].count()[0] #num of fraud records
y = crd_df[crd_df.Class == 0 ].count()[0] #num of non-fraud records
z = x/(x+y) #proportion of fraud records
print('In our sample the proportion of fraud is:',round(z*100,3),'%.')

In our sample the proportion of fraud is: 0.173 %.


Since the dataset states that only 0.173% of the sample are frauds, we will get the top 0.173% of our sample and set the prediction to True for these values. For the rest, we will set the prediction of our model to False. Finally we will calculate the accuracy.

Note: Another approach would be to classify according to probability :
crd_df.Prediction >= 0.173% => fraud 
else not fraud.

The method we chose is based on the ranking of the probabilities and not the absolute value, since we know 0.173% is fraud. For example, if our model underestimated the probability, we might ended up getting a lower percentage of frauds than the actual 0.173% we know that is the mean of our population. The same would apply if our model overestimated  the probability value.

In [7]:
limit = crd_df.Prediction.quantile(1-z) #we limit our prediction to top (1-0.00173) quantile
crd_df ['C_Pred'] = np.where(crd_df.Prediction >= limit, True, False) #and set them as fraud predictions

We will split into two dataframes, 'positives' and 'negatives' for positive and negative predictions, respectively.

In [8]:
negatives = crd_df[crd_df.C_Pred == False] #negative predictions 
positives = crd_df[crd_df.C_Pred == True] #positive predictions

In [9]:
fn = negatives[negatives.Class == 1 ].count()[0] #false negatives
tn = negatives[negatives.Class == 0 ].count()[0] #true negatives
print('False negatives are',fn,'and true negatives are',tn,'.')

False negatives are 103 and true negatives are 284212 .


In [10]:
fp = positives[positives.Class == 0 ].count()[0] #false positives
tp = positives[positives.Class == 1 ].count()[0] #true positives
print('False positives are',fp,'and true positives are',tp,'.')

False positives are 103 and true positives are 389 .


In [11]:
accuracy = (tn+tp)/(fn+tn+fp+tp) #total accuracy
print('Accuracy measured as the ratio of true predictions over total predictions is:',round(accuracy*100,2),'%.')

Accuracy measured as the ratio of true predictions over total predictions is: 99.93 %.


In [12]:
precision = tp/(tp+fp) #precision is percentage of true positives over total positives predicted
print('The precision is',round(precision*100,2),'%.')

The precision is 79.07 %.


In [13]:
recall = tp/(tp+fn) #recall is percentage of true positives over all actual positives
print('The recall is',round(recall*100,2),'%.')

The recall is 79.07 %.


Now, we would to improve our model so that it is more accurate, and have higher precision and recall. 

One way is to undersample our data. We will split data into two categories according to the amount of the transaction and check whether the fraud rates remains around 0,172% or if it differs.

One hypothesis is that fraud could be much more common for small amounts. Of course the same applies for large amounts. The whole point is that fraud could be detected with different possibilities in distinct subsets of our data.

Since we have no additional information we will try many values for splitting data according to amount quantiles, until we observe a difference in the proportion of fraud ( if we finally do). We decide to check at the tails to see any difference, because usually in statistics they are areas of great interest and may hide extra information.

In [14]:
test_limit = [0.005,0.01,0.05,0.10,0.25,0.50,0.75,0.90,0.95,0.99,0.995,0.999]

for i in test_limit:
    lim = crd_df.Amount.quantile(i)
    cm_crd_df = crd_df[crd_df.Amount > lim ].copy() #copy more than limit
    x_1 = cm_crd_df[cm_crd_df.Class == 0].count()[0] 
    y_1 = cm_crd_df[cm_crd_df.Class == 1].count()[0]
    z_1 = y_1/(x_1+y_1)
    cl_crd_df = crd_df[crd_df.Amount <= lim ].copy() #copy less than limit   
    x_2 = cl_crd_df[cl_crd_df.Class == 0].count()[0]
    y_2 = cl_crd_df[cl_crd_df.Class == 1].count()[0]
    z_2 = y_2/(x_2+y_2)
    print('Fraud for amounts greater than', round(lim,2), ', the proportion of fraud is',round(z_1*100,2),'%.')
    print('For amounts smaller or equal to',round(lim,2),',the proportion of fraud is',round(z_2*100,2),'%.')
    print('Respective quantile point is',i,'.')
    print()

Fraud for amounts greater than 0.0 , the proportion of fraud is 0.16 %.
For amounts smaller or equal to 0.0 ,the proportion of fraud is 1.48 %.
Respective quantile point is 0.005 .

Fraud for amounts greater than 0.12 , the proportion of fraud is 0.16 %.
For amounts smaller or equal to 0.12 ,the proportion of fraud is 1.07 %.
Respective quantile point is 0.01 .

Fraud for amounts greater than 0.92 , the proportion of fraud is 0.16 %.
For amounts smaller or equal to 0.92 ,the proportion of fraud is 0.47 %.
Respective quantile point is 0.05 .

Fraud for amounts greater than 1.0 , the proportion of fraud is 0.12 %.
For amounts smaller or equal to 1.0 ,the proportion of fraud is 0.59 %.
Respective quantile point is 0.1 .

Fraud for amounts greater than 5.6 , the proportion of fraud is 0.13 %.
For amounts smaller or equal to 5.6 ,the proportion of fraud is 0.31 %.
Respective quantile point is 0.25 .

Fraud for amounts greater than 22.0 , the proportion of fraud is 0.16 %.
For amounts smalle

Normally, fraud should be around 0.17%.

We notice that for amounts greater than 203.0 , which corresponds to 0.90 quantile, fraud is found at 0.3% of the data, which is almost double as expected. 

Thus, we will split the data into the 0.95 quantile and perform analysis seperately at the two sets, following the same strategy with the initial, in order to compare the results.

Of course, we could split our data into many more sets, achieving higher accuracy but then we would overfit our sample data and probably would have lower accuracy on the population.

In [15]:
lim_final = crd_df.Amount.quantile(0.9) #get 0.9 quantile value
cm_crd_df = crd_df[crd_df.Amount > lim_final ].copy() #first subset
cl_crd_df = crd_df[crd_df.Amount <= lim_final ].copy() #second subset

We perform logistic regression at the two datasets independently.

In [16]:
f_m = 'Class ~ Time+V1+V2+V3+V4+V5+V6+V7+V8+V9+V10+V11+V12+V13+V14+V15+V16+V17+V18+V19+V20+V21+V22+V23+V24+V25+V26+V27+V28+Amount'
logitfit_m = smf.logit(formula = str(f_m), data = cm_crd_df).fit() #fit our model

Optimization terminated successfully.
         Current function value: 0.005169
         Iterations 13


In [17]:
f_l = 'Class ~ Time+V1+V2+V3+V4+V5+V6+V7+V8+V9+V10+V11+V12+V13+V14+V15+V16+V17+V18+V19+V20+V21+V22+V23+V24+V25+V26+V27+V28+Amount'
logitfit_l = smf.logit(formula = str(f_l), data = cl_crd_df).fit() #fit our model

Optimization terminated successfully.
         Current function value: 0.003434
         Iterations 13


In [18]:
logitfit_m.summary() #summay for first model

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  Class   No. Observations:                28458
Model:                          Logit   Df Residuals:                    28427
Method:                           MLE   Df Model:                           30
Date:                Sun, 14 Jan 2018   Pseudo R-squ.:                  0.7460
Time:                        20:19:25   Log-Likelihood:                -147.10
converged:                       True   LL-Null:                       -579.02
                                        LLR p-value:                2.443e-162
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -8.4722      0.806    -10.511      0.000     -10.052      -6.892
Time       -1.165e-05   6.86e-06     -1.698      0.090   -2.51e-05     1.8e-06
V1             0.8595      0.401      2.146      0.032       0.074       1.645
V2            -0.3104      0.356     -0.871      0.384      -1.009       0.388
V3             0.3035      0.274      1.108      0.268      -0.233       0.840
V4             1.1586      0.267      4.341      0.000       0.635       1.682
V5             0.4749      0.272      1.744      0.081      -0.059       1.009
V6            -0.4778      0.309     -1.548      0.122      -1.083       0.127
V7            -0.0143      0.238     -0.060      0.952      -0.480       0.452
V8            -0.5254      0.203     -2.584      0.010      -0.924      -0.127
V9            -0.9914      0.507     -1.955      0.051      -1.985       0.003
V10           -2.3631      0.721     -3.276      0.001      -3.777      -0.949
V11           -0.1569      0.239     -0.656      0.512      -0.626       0.312
V12           -0.6723      0.232     -2.903      0.004      -1.126      -0.218
V13           -0.3321      0.210     -1.585      0.113      -0.743       0.079
V14           -0.1839      0.199     -0.925      0.355      -0.574       0.206
V15            0.0789      0.252      0.314      0.754      -0.414       0.572
V16           -0.3830      0.302     -1.268      0.205      -0.975       0.209
V17           -0.0574      0.194     -0.297      0.767      -0.437       0.322
V18            0.1553      0.343      0.453      0.651      -0.517       0.828
V19            0.4752      0.308      1.545      0.122      -0.127       1.078
V20           -1.5223      0.578     -2.634      0.008      -2.655      -0.389
V21            0.5273      0.245      2.149      0.032       0.046       1.008
V22            0.4915      0.347      1.418      0.156      -0.188       1.171
V23            0.0826      0.178      0.465      0.642      -0.265       0.431
V24            0.1784      0.496      0.360      0.719      -0.794       1.151
V25           -0.4573      0.424     -1.078      0.281      -1.288       0.374
V26           -1.3171      0.535     -2.462      0.014      -2.366      -0.268
V27           -2.1465      0.844     -2.543      0.011      -3.801      -0.492
V28           -0.6290      0.361     -1.744      0.081      -1.336       0.078
Amount         0.0022      0.002      1.283      0.200      -0.001       0.006
==============================================================================

Possibly complete quasi-separation: A fraction 0.48 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

In [19]:
logitfit_l.summary() #summary for second model

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  Class   No. Observations:               256349
Model:                          Logit   Df Residuals:                   256318
Method:                           MLE   Df Model:                           30
Date:                Sun, 14 Jan 2018   Pseudo R-squ.:                  0.7095
Time:                        20:19:31   Log-Likelihood:                -880.18
converged:                       True   LL-Null:                       -3030.0
                                        LLR p-value:                     0.000
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -8.3862      0.288    -29.086      0.000      -8.951      -7.821
Time       -3.483e-06   2.53e-06     -1.378      0.168   -8.44e-06    1.47e-06
V1             0.0764      0.045      1.686      0.092      -0.012       0.165
V2            -0.0215      0.067     -0.323      0.747      -0.152       0.109
V3            -0.0233      0.059     -0.394      0.693      -0.139       0.092
V4             0.6406      0.081      7.890      0.000       0.481       0.800
V5             0.0818      0.072      1.144      0.253      -0.058       0.222
V6            -0.0582      0.077     -0.751      0.453      -0.210       0.094
V7            -0.2038      0.071     -2.852      0.004      -0.344      -0.064
V8            -0.1658      0.034     -4.830      0.000      -0.233      -0.098
V9            -0.1896      0.124     -1.533      0.125      -0.432       0.053
V10           -0.5597      0.107     -5.219      0.000      -0.770      -0.350
V11           -0.0392      0.090     -0.433      0.665      -0.216       0.138
V12            0.1590      0.097      1.642      0.101      -0.031       0.349
V13           -0.3171      0.092     -3.449      0.001      -0.497      -0.137
V14           -0.7046      0.071     -9.903      0.000      -0.844      -0.565
V15           -0.1395      0.097     -1.442      0.149      -0.329       0.050
V16           -0.2043      0.144     -1.415      0.157      -0.487       0.079
V17           -0.0205      0.077     -0.265      0.791      -0.172       0.131
V18            0.0259      0.147      0.176      0.860      -0.263       0.315
V19           -0.0260      0.109     -0.240      0.811      -0.239       0.187
V20           -0.3784      0.110     -3.440      0.001      -0.594      -0.163
V21            0.3798      0.075      5.088      0.000       0.234       0.526
V22            0.7771      0.159      4.880      0.000       0.465       1.089
V23           -0.1769      0.065     -2.741      0.006      -0.303      -0.050
V24            0.1229      0.161      0.765      0.444      -0.192       0.438
V25            0.0317      0.151      0.210      0.834      -0.264       0.327
V26            0.2537      0.215      1.179      0.239      -0.168       0.676
V27           -0.8950      0.135     -6.633      0.000      -1.159      -0.631
V28           -0.3502      0.107     -3.261      0.001      -0.561      -0.140
Amount        -0.0058      0.002     -2.406      0.016      -0.010      -0.001
==============================================================================

Possibly complete quasi-separation: A fraction 0.40 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

Now, we will caclulate the possibilies for the records to be classified as frauds.

In [20]:
y_pred_l = logitfit_l.predict() #model predict
cl_crd_df['Prediction'] = (y_pred_l) #store result
cl_crd_df.head(10) #select first 10 records

Time        V1        V2        V3        V4        V5        V6  \
0    0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1    0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
3    1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4    2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
5    2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728   
6    4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708   
7    7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118   
8    7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818   
9    9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761   
10  10.0  1.449044 -1.176339  0.913860 -1.375667 -1.971383 -0.629152   

          V7        V8        V9   ...         V23       V24       V25  \
0   0.239599  0.098698  0.363787   ...   -0.110474  0.066928  0.128539   
1  -0.078803  0.085102 -0.255425   ...    0.101288 -0.339846  0.167170   
3   0.237609  0.377436 -1.387024   ...   -0.190321 -1.175575  0.647376   
4   0.592941 -0.270533  0.817739   ...   -0.137458  0.141267 -0.206010   
5   0.476201  0.260314 -0.568671   ...   -0.026398 -0.371427 -0.232794   
6  -0.005159  0.081213  0.464960   ...   -0.154104 -0.780055  0.750137   
7   1.120631 -3.807864  0.615375   ...    0.057504 -0.649709 -0.415267   
8   0.370145  0.851084 -0.392048   ...   -0.204233  1.011592  0.373205   
9   0.651583  0.069539 -0.736727   ...   -0.120794 -0.385050 -0.069733   
10 -1.423236  0.048456 -1.720408   ...    0.027740  0.500512  0.251367   

         V26       V27       V28  Amount  Class  Prediction  C_Pred  
0  -0.189115  0.133558 -0.021053  149.62      0    0.000214   False  
1   0.125895 -0.008983  0.014724    2.69      0    0.000187   False  
3  -0.221929  0.062723  0.061458  123.50      0    0.000086   False  
4   0.502292  0.219422  0.215153   69.99      0    0.000207   False  
5   0.105915  0.253844  0.081080    3.67      0    0.000107   False  
6  -0.257237  0.034507  0.005168    4.99      0    0.000447   False  
7  -0.051634 -1.206921 -1.085339   40.80      0    0.000463   False  
8  -0.384157  0.011747  0.142404   93.20      0    0.000102   False  
9   0.094199  0.246219  0.083076    3.68      0    0.000086   False  
10 -0.129478  0.042850  0.016253    7.80      0    0.000102   False  

[10 rows x 33 columns]

In [21]:
y_pred_m = logitfit_m.predict() #model predict
cm_crd_df['Prediction'] = (y_pred_m) #store result
cm_crd_df.head(10) #select first 10 records

Time        V1        V2        V3        V4        V5        V6  \
2     1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
20   16.0  0.694885 -1.361819  1.029221  0.834159 -1.191209  1.309109   
51   36.0 -1.004929 -0.985978 -0.038039  3.710061 -6.631951  5.122103   
64   42.0 -0.522666  1.009923  0.276470  1.475289 -0.707013  0.355243   
89   59.0 -0.773293 -4.146007 -0.932038  0.027094 -1.698307  0.460188   
134  83.0 -1.897331  0.955626  0.052543  1.276656 -3.323084  3.229911   
140  87.0 -5.101877  1.897022 -3.458034 -1.277543 -5.517758  2.098366   
142  89.0  0.270725 -1.615317  1.054982  1.661510 -1.737687  0.065894   
147  92.0  0.607153 -0.957413  0.761505  1.590050 -1.177298  0.011232   
149  93.0 -0.853807  0.127392  1.267277  0.678584 -1.029851 -0.487614   

           V7        V8        V9   ...         V23       V24       V25  \
2    0.791461  0.247676 -1.514654   ...    0.909412 -0.689281 -0.327642   
20  -0.878586  0.445290 -0.446196   ...   -0.050881 -0.304215  0.072001   
51   4.371691 -2.006868 -0.278736   ...    0.969719  0.019445  0.570923   
64   1.559849 -0.399579 -0.479813   ...    0.069666  0.157820 -1.109224   
89   0.737344 -0.314216 -0.842673   ...   -1.134769 -0.654958  0.098386   
134  1.029631  1.515607 -0.059627   ...    0.477537 -0.608981 -1.120892   
140  3.329603  1.250966  0.271501   ...   -0.555900 -0.761660  0.066611   
142 -0.313977  0.089081  1.069842   ...   -0.482194  0.418871  0.235961   
147 -0.166991  0.152720  0.742860   ...   -0.317001  0.543778  0.484232   
149  1.836071 -0.298566 -0.922127   ...    0.476744  0.724705  0.526798   

          V26       V27       V28   Amount  Class    Prediction  C_Pred  
2   -0.139097 -0.055353 -0.059752   378.66      0  4.548700e-04   False  
20  -0.422234  0.086553  0.063499   231.71      0  1.372466e-04   False  
51   0.333278  0.857373 -0.075538  1402.95      0  1.511047e-03   False  
64  -0.302369  0.318170  0.316910   243.66      0  6.623480e-04   False  
89  -0.209150 -0.171709  0.208057  1142.02      0  7.629922e-05   False  
134 -0.413851  0.061399 -0.187964   552.18      0  1.831699e-03   False  
140  0.767227  0.731634 -0.860310   919.60      0  1.721692e-09   False  
142 -0.265185 -0.001063  0.120126   459.39      0  2.316215e-03   False  
147 -0.330425 -0.001545  0.062098   268.78      0  2.134259e-03   False  
149  0.502701 -0.159465  0.002761   322.44      0  1.987212e-03   False  

[10 rows x 33 columns]

Now, we will proceed evaluating our model compared to previous one.

In [22]:
x_l = cl_crd_df[cl_crd_df.Class == 1 ].count()[0] #number of fraud records in first dataframe
y_l = cl_crd_df[cl_crd_df.Class == 0 ].count()[0] #number of non-fraud records in first dataframe
z_l = x_l/(x_l+y_l) #percentage of fraud records in first dataframe

x_m = cm_crd_df[cm_crd_df.Class == 1 ].count()[0] #number of fraud records in second dataframe
y_m = cm_crd_df[cm_crd_df.Class == 0 ].count()[0] #number of non-fraud records in second dataframe
z_m = x_m/(x_m+y_m)#percentage of fraud records in second dataframe

print('In the sample with the lower amounts, the percentage of fraud is:',round(z_l*100,3),'%.')
print('In the sample with the higher amounts, the percentage of fraud is:',round(z_m*100,3),'%.')

In the sample with the lower amounts, the percentage of fraud is: 0.159 %.
In the sample with the higher amounts, the percentage of fraud is: 0.299 %.


We apply the same strategy as previously based on quantiles (ranking) for each dataframe.

In [23]:
limit_l = cl_crd_df.Prediction.quantile(1-z_l) 
cl_crd_df ['C_Pred'] = np.where(cl_crd_df.Prediction >= limit_l, True, False)
cl_crd_df.head(10)

Time        V1        V2        V3        V4        V5        V6  \
0    0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1    0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
3    1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4    2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
5    2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728   
6    4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708   
7    7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118   
8    7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818   
9    9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761   
10  10.0  1.449044 -1.176339  0.913860 -1.375667 -1.971383 -0.629152   

          V7        V8        V9   ...         V23       V24       V25  \
0   0.239599  0.098698  0.363787   ...   -0.110474  0.066928  0.128539   
1  -0.078803  0.085102 -0.255425   ...    0.101288 -0.339846  0.167170   
3   0.237609  0.377436 -1.387024   ...   -0.190321 -1.175575  0.647376   
4   0.592941 -0.270533  0.817739   ...   -0.137458  0.141267 -0.206010   
5   0.476201  0.260314 -0.568671   ...   -0.026398 -0.371427 -0.232794   
6  -0.005159  0.081213  0.464960   ...   -0.154104 -0.780055  0.750137   
7   1.120631 -3.807864  0.615375   ...    0.057504 -0.649709 -0.415267   
8   0.370145  0.851084 -0.392048   ...   -0.204233  1.011592  0.373205   
9   0.651583  0.069539 -0.736727   ...   -0.120794 -0.385050 -0.069733   
10 -1.423236  0.048456 -1.720408   ...    0.027740  0.500512  0.251367   

         V26       V27       V28  Amount  Class  Prediction  C_Pred  
0  -0.189115  0.133558 -0.021053  149.62      0    0.000214   False  
1   0.125895 -0.008983  0.014724    2.69      0    0.000187   False  
3  -0.221929  0.062723  0.061458  123.50      0    0.000086   False  
4   0.502292  0.219422  0.215153   69.99      0    0.000207   False  
5   0.105915  0.253844  0.081080    3.67      0    0.000107   False  
6  -0.257237  0.034507  0.005168    4.99      0    0.000447   False  
7  -0.051634 -1.206921 -1.085339   40.80      0    0.000463   False  
8  -0.384157  0.011747  0.142404   93.20      0    0.000102   False  
9   0.094199  0.246219  0.083076    3.68      0    0.000086   False  
10 -0.129478  0.042850  0.016253    7.80      0    0.000102   False  

[10 rows x 33 columns]

In [24]:
limit_m = cm_crd_df.Prediction.quantile(1-z_m)
cm_crd_df ['C_Pred'] = np.where(cm_crd_df.Prediction >= limit_m, True, False)
cm_crd_df.head(10)

Time        V1        V2        V3        V4        V5        V6  \
2     1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
20   16.0  0.694885 -1.361819  1.029221  0.834159 -1.191209  1.309109   
51   36.0 -1.004929 -0.985978 -0.038039  3.710061 -6.631951  5.122103   
64   42.0 -0.522666  1.009923  0.276470  1.475289 -0.707013  0.355243   
89   59.0 -0.773293 -4.146007 -0.932038  0.027094 -1.698307  0.460188   
134  83.0 -1.897331  0.955626  0.052543  1.276656 -3.323084  3.229911   
140  87.0 -5.101877  1.897022 -3.458034 -1.277543 -5.517758  2.098366   
142  89.0  0.270725 -1.615317  1.054982  1.661510 -1.737687  0.065894   
147  92.0  0.607153 -0.957413  0.761505  1.590050 -1.177298  0.011232   
149  93.0 -0.853807  0.127392  1.267277  0.678584 -1.029851 -0.487614   

           V7        V8        V9   ...         V23       V24       V25  \
2    0.791461  0.247676 -1.514654   ...    0.909412 -0.689281 -0.327642   
20  -0.878586  0.445290 -0.446196   ...   -0.050881 -0.304215  0.072001   
51   4.371691 -2.006868 -0.278736   ...    0.969719  0.019445  0.570923   
64   1.559849 -0.399579 -0.479813   ...    0.069666  0.157820 -1.109224   
89   0.737344 -0.314216 -0.842673   ...   -1.134769 -0.654958  0.098386   
134  1.029631  1.515607 -0.059627   ...    0.477537 -0.608981 -1.120892   
140  3.329603  1.250966  0.271501   ...   -0.555900 -0.761660  0.066611   
142 -0.313977  0.089081  1.069842   ...   -0.482194  0.418871  0.235961   
147 -0.166991  0.152720  0.742860   ...   -0.317001  0.543778  0.484232   
149  1.836071 -0.298566 -0.922127   ...    0.476744  0.724705  0.526798   

          V26       V27       V28   Amount  Class    Prediction  C_Pred  
2   -0.139097 -0.055353 -0.059752   378.66      0  4.548700e-04   False  
20  -0.422234  0.086553  0.063499   231.71      0  1.372466e-04   False  
51   0.333278  0.857373 -0.075538  1402.95      0  1.511047e-03   False  
64  -0.302369  0.318170  0.316910   243.66      0  6.623480e-04   False  
89  -0.209150 -0.171709  0.208057  1142.02      0  7.629922e-05   False  
134 -0.413851  0.061399 -0.187964   552.18      0  1.831699e-03   False  
140  0.767227  0.731634 -0.860310   919.60      0  1.721692e-09   False  
142 -0.265185 -0.001063  0.120126   459.39      0  2.316215e-03   False  
147 -0.330425 -0.001545  0.062098   268.78      0  2.134259e-03   False  
149  0.502701 -0.159465  0.002761   322.44      0  1.987212e-03   False  

[10 rows x 33 columns]

Now, we will split the two datasets into positive and negative predictions, respectively.

In [25]:
negatives_l = cl_crd_df[cl_crd_df.C_Pred == False]
positives_l = cl_crd_df[cl_crd_df.C_Pred == True]
negatives_m = cm_crd_df[cm_crd_df.C_Pred == False]
positives_m = cm_crd_df[cm_crd_df.C_Pred == True]

In [26]:
fn_l = negatives_l[negatives_l.Class == 1 ].count()[0]
tn_l = negatives_l[negatives_l.Class == 0 ].count()[0]
print('False negatives in lower amounts dataset are',fn_l,'and true negatives are',tn_l,'.')

False negatives in lower amounts dataset are 91 and true negatives are 255851 .


In [27]:
fn_m = negatives_m[negatives_m.Class == 1 ].count()[0]
tn_m = negatives_m[negatives_m.Class == 0 ].count()[0]
print('False negatives in lower amounts dataset are',fn_m,'and true negatives are',tn_m,'.')

False negatives in lower amounts dataset are 22 and true negatives are 28351 .


In [28]:
fp_l = positives_l[positives_l.Class == 0 ].count()[0]
tp_l = positives_l[positives_l.Class == 1 ].count()[0]
print('False positives in higher amounts dataset are',fp_l,'and true positives are',tp_l,'.')

False positives in higher amounts dataset are 91 and true positives are 316 .


In [29]:
fp_m = positives_m[positives_m.Class == 0 ].count()[0]
tp_m = positives_m[positives_m.Class == 1 ].count()[0]
print('False positives in higher amounts dataset are',fp_m,'and true positives are',tp_m,'.')

False positives in higher amounts dataset are 22 and true positives are 63 .


In [30]:
new_accuracy = (tn_l+tn_m+tp_l+tp_m)/(fn_l+fn_m+tn_l+tn_m+fp_l+fp_m+tp_l+tp_m)
print('New accuracy  is:',round(new_accuracy*100,2),'%.')

New accuracy  is: 99.92 %.


In [31]:
new_precision = (tp_l+tp_m)/(tp_l+tp_m+fp_l+fp_m)
print('The precision is',round(new_precision*100,2),'%.')

The precision is 77.03 %.


In [32]:
new_recall = (tp_l+tp_m)/(tp_l+tp_m+fn_l+fn_m)
print('The recall is',round(new_recall*100,2),'%.')

The recall is 77.03 %.


Undersampling our initial sample was a promising strategy, but finally it did not work as expected. 

This is a kind of indication that when we have to deal with highly imbalancing data, we have to test more than one options and check what works well. 

More advanced algorithms (xgboost - decision trees and undersampling) would might improve ratio and precision and total accuracy of our predictions. Hopefully, we would might achieve better accuracy results, probably resulting in parameter tuning and complexity time increasing.